eddymotion/viz.py

```diff
@@ -35,3 +35,159 @@ def plot_dwi(dataobj, affine, gradient=None, **kwargs):
 $\\vec{{b}}$ = ({', '.join(str(v) for v in gradient[:3])})""",
         **kwargs,
     )
+
+
+_epsi = 1.0e-9
+
+
+def rotation_matrix(u, v):
+    """
+    Returns a rotation matrix R s.t. Ru = v.
+
+    Extracted from the code of Emmanuel Caruyer
+    (https://github.com/ecaruyer/qspace/blob/master/qspace/visu/visu_points.py).
+    """
+    # the axis is given by the product u x v
+    u = u / np.sqrt((u ** 2).sum())
+    v = v / np.sqrt((v ** 2).sum())
+    w = np.asarray([u[1] * v[2] - u[2] * v[1],
+                    u[2] * v[0] - u[0] * v[2],
+                    u[0] * v[1] - u[1] * v[0]])
+    if (w ** 2).sum() < _epsi:
+        # The vectors u and v are collinear
+        return np.eye(3)
+
+    # computes sine and cosine
+    c = np.dot(u, v)
+    s = np.sqrt((w ** 2).sum())
+
+    w = w / s
+    P = np.outer(w, w)
+    Q = np.asarray([[0, -w[2], w[1]], [w[2], 0, -w[0]], [-w[1], w[0], 0]])
+    R = P + c * (np.eye(3) - P) + s * Q
+    return R
+
+
+def draw_circles(positions, rs):
+    """
+    Draw circular patches (lying on a sphere) at given positions.
+
+    Adapted from the code of Emmanuel Caruyer
+    (https://github.com/ecaruyer/qspace/blob/master/qspace/visu/visu_points.py)
+    by the Nipreps developers.
+    """
+    # a circle centered at [1, 0, 0] with radius r
+    M = 20
+    t = np.linspace(0, 2 * np.pi, M)
+
+    nbPoints = positions.shape[0]
+    circles = np.zeros((nbPoints, M, 3))
+    for i in range(positions.shape[0]):
+        circleX = np.zeros((20, 3))
+        dots_radius = np.sqrt(rs[i]) * 0.04
+        circleX[:, 1] = dots_radius * np.cos(t)
+        circleX[:, 2] = dots_radius * np.sin(t)
+        norm = np.sqrt((positions[i] ** 2).sum())
+        point = positions[i] / norm
+        R1 = rotation_matrix(np.asarray([1, 0, 0]), point)
+        circles[i] = positions[i] + np.dot(R1, circleX.T).T
+    return circles
+
+
+def draw_points(gradients, ax, colormap='viridis'):
+    """
+    Draw the vectors on a shell.
+
+    Adapted from the code of Emmanuel Caruyer
+    (https://github.com/ecaruyer/qspace/blob/master/qspace/visu/visu_points.py)
```

Comment:
I think that the way to copy this code over is to include the [original license](https://github.com/ecaruyer/qspace/blob/master/LICENSE) somewhere in this file (you can also additionally include a note in every docstring pointing to the original implementation).

Reply:
I *believe* the BSD license does not mandate that (although, I agree what you propose would be the right thing to do).

I propose:

1. Retain the copyright notice at the top of the file Ariel liked, as mandated for redistribution of the original source code by BSD.
2. Make a list of the most relevant modifications of the code, as if this were some sort of Apache License.

Reply:
I'll send a proposal of how to do this as soon as I get the time.

Reply:
Thanks Oscar! It is indeed not clear yet to me how to make it the most clear. 
In summary:
* the method `rotation_matrix` is exactly the code of Emmanuel.
* the method `draw_circle` has been modified to take the full list of normalized bvecs and corresponding circle radii instead of taking the list of bvecs and radii for a specific shell (bvalue)
* the methods `draw_points` has been majorly modified including:
  * the input is a single 2D numpy array of the gradient table in RAS+B format
  * the scaling of the circle radius for each bvec proportional to the inverse of the bvals. A minimum/maximal value for the radii can be specified
  *  Circles for each bvec are drawn at once instead of looping over the shells.
  * Some variables have been renamed (like `vects` to `bvecs`)